test/goalc/test_arithmetic.cpp

```diff
@@ -0,0 +1,232 @@
+    }
+    return std::to_string(val);
+  }
+
+	// This is used by the test runner code to know what the expected value is
+	// For a simple example like this, a single eval is all that's required, but for
+	// more complex tests, this may not be the case.
+  std::string eval() {
+    if (hex) {
+      int64_t hexVal;
+      std::stringstream ss;
+      ss << std::hex << std::to_string(val);
+      ss >> hexVal;
+      return std::string(std::to_string(hexVal)) + "\n";
+    }
+    return std::to_string(val) + "\n";
+  }
+};
+
+// This is a helper function that is used to generate a bunch of tests
+// once again, very over-engineered for just testing engineers, but you might imagine
+// a more complex test template that has several conditionals / loops / etc
+std::vector<IntegerParam> genIntegerTests(int numTests,
+                                          bool includeHex,
+                                          bool includeNegative,
+                                          std::vector<IntegerParam> additionalTests = {}) {
+  std::vector<IntegerParam> tests;
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<std::mt19937::result_type> dist6(0, UINT32_MAX);
+  int testCases = includeNegative ? 2 : 1;
+  if (includeHex) {
+    testCases *= 2;
+  }
+  for (int i = 0; i < numTests; i++) {
+    switch (i % testCases) {
+      case 0:
+        tests.push_back(IntegerParam(dist6(rng), false, i));
+        break;
+      case 1:
+        tests.push_back(IntegerParam(dist6(rng) * -1, false, i));
+        break;
+      case 2:
+        tests.push_back(IntegerParam(dist6(rng), true, i));
+        break;
+      case 3:
+        tests.push_back(IntegerParam(dist6(rng) * -1, true, i));
+        break;
+    }
+  }
+
+  for (int i = 0; i < additionalTests.size(); i++) {
+    IntegerParam test = additionalTests.at(i);
+    test.index = i + numTests - 1;
+    tests.push_back(test);
+  }
+
+  return tests;
+}
+
+// In the interest of speed, we want to share the same thread/compiler across
+// all the tests in this suite, so we have to over-ride this.
+class ArithmeticTests : public testing::TestWithParam<IntegerParam> {
+ public:
+  // Per-test-suite set-up.
+  // Called before the first test in this test suite.
+  static void SetUpTestSuite() {
+    runtime_thread = std::thread((GoalTest::runtime_no_kernel));
+    runner.c = &compiler;
+  }
+
+  // Per-test-suite tear-down.
+  // Called after the last test in this test suite.
+  static void TearDownTestSuite() {
+    compiler.shutdown_target();
+    runtime_thread.join();
+  }
+
+  // You can define per-test set-up logic as usual.
+  void SetUp() {
+    GoalTest::createDirIfAbsent(GoalTest::getTemplateDir(testCategory));
+    GoalTest::createDirIfAbsent(GoalTest::getGeneratedDir(testCategory));
+  }
+
+  // You can define per-test tear-down logic as usual.
+  void TearDown() {}
+
+  // Common Resources Across all Tests in the Suite
+  static std::thread runtime_thread;
```

Comment:
This seems kinda weird, but maybe this is what you're supposed to do for gtests?

Reply:
Well, using gtest's Fixture support was the best way I could think of to share the thread/compiler/runner across all tests.

However I don't really like it for several reasons:
- Since the Fixture has to extend with the parameter, you can't easily have 2 different types of parameters to generate tests, unless you make a new fixture, which means a new thread,etc.
- Lots of boilerplate

So this is an area for improvement.  I think it would be nicer to have a "global" runner with/without the kernel, and have some sort of mechanism to "refresh/wipe" the state so we can ensure tests are hermetic.  I ran into some difficulty trying to do the former though.

There might also be a better way to do the parameterized tests - https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#creating-value-parameterized-abstract-tests so we can have a common library of some sorts.

Reply:
I looked at this a little bit more and I think what you have now is good. The `static` stuff in the fixture class seemed weird to me at first because I didn't realize each test would get its own fixture.

I agree we need a better way to reset the state of the compiler/runtime between tests. I don't think it would be too hard to add a `reset_state` to the compiler that clears out all the variable/constant/function/type definitions.  Doing a fast reset of the runtime is going to be much harder though.  There's some hacky stuff we could do like saving the entire runtime GOAL memory before a test and restoring it, but this won't catch everything.